compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -1077,6 +1049,211 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 ),
             );
         }
+    }
+
+    fn report_no_match_method_error(
+        &self,
+        mut span: Span,
+        rcvr_ty: Ty<'tcx>,
+        item_ident: Ident,
+        expr_id: hir::HirId,
+        source: SelfSource<'tcx>,
+        args: Option<&'tcx [hir::Expr<'tcx>]>,
+        sugg_span: Span,
+        no_match_data: &mut NoMatchData<'tcx>,
+        expected: Expectation<'tcx>,
+        trait_missing_method: bool,
+        within_macro_span: Option<Span>,
+    ) -> ErrorGuaranteed {
+        let mode = no_match_data.mode;
+        let tcx = self.tcx;
+        let rcvr_ty = self.resolve_vars_if_possible(rcvr_ty);
+        let mut ty_file = None;
+        let is_method = mode == Mode::MethodCall;
+        let unsatisfied_predicates = &no_match_data.unsatisfied_predicates;
+        let similar_candidate = no_match_data.similar_candidate;
+        let item_kind = if is_method {
+            "method"
+        } else if rcvr_ty.is_enum() {
+            "variant or associated item"
+        } else {
+            match (item_ident.as_str().chars().next(), rcvr_ty.is_fresh_ty()) {
+                (Some(name), false) if name.is_lowercase() => "function or associated item",
+                (Some(_), false) => "associated item",
+                (Some(_), true) | (None, false) => "variant or associated item",
+                (None, true) => "variant",
+            }
+        };
+
+        // We could pass the file for long types into these two, but it isn't strictly necessary
+        // given how targeted they are.
+        if let Err(guar) =
+            self.report_failed_method_call_on_range_end(tcx, rcvr_ty, source, span, item_ident)
+        {
+            return guar;
+        }
+        if let Err(guar) = self.report_failed_method_call_on_numerical_infer_var(
+            tcx,
+            rcvr_ty,
+            source,
+            span,
+            item_kind,
+            item_ident,
+            &mut ty_file,
+        ) {
+            return guar;
+        }
+        span = item_ident.span;
```

Comment:
that's weird. why? and could we just use `item_ident.span` directly below this point?